kernelCTF: Add CVE-2024-41010_lts (#126)
google/security-research: pocs/linux/kernelctf/CVE-2024-41010_lts/docs/exploit.md

# Exploit detail about CVE-2024-41010

## Primitive

> When UAF triggered at `uaf_chunk`, we can modify `uaf_chunk[0][0x20, 0x28)` to `start_poll_synchronize_rcu()` return value.

Reviewer: Thanks for clarifying this in recent `exploit.md` update. Closing the comment.

	/* size: 40, cachelines: 1, members: 4 */
	/* forced alignments: 1 */
	/* last cacheline: 40 bytes */
} __attribute__((__aligned__(8)));
```

So, by using Cross-Cache, we overlap `struct simple_xattr` with `uaf_chunk`

### Modify `struct simple_xattr`'s `size`
At [Trigger Vulnerability](#trigger-vulnerability), we know that can modify `uaf_chunk[0][0x20, 0x28)` to `start_poll_synchronize_rcu()` return value. Therefore, by overlapping, we can control `struct simple_xattr`'s `size` as `start_poll_synchronize_rcu()` return value.

The import things is that the first value of `struct simple_xattr` is `rb_node`. The `struct rb_node` is as follows:
```c++
struct rb_node {
	long unsigned int          __rb_parent_color;    /*     0     8 */
	struct rb_node *           rb_right;             /*     8     8 */
	struct rb_node *           rb_left;              /*    16     8 */

	/* size: 24, cachelines: 1, members: 3 */
	/* last cacheline: 24 bytes */
} __attribute__((__aligned__(8)));
```

Therefore, `uaf_chunk[0]` is `(struct rb_node *)uaf_chunk -> __rb_parent_color`.  `__rb_parent_color` is constructed as follows:
```c++
#define RB_RED          0
#define RB_BLACK        1

> __rb_parent_color = (unsigned long)parent | color;

Reviewer: Could you please expand code block here, so that it's clear that parent is a pointer, like this: 

```
static inline void rb_link_node(struct rb_node *node, struct rb_node *parent,
				struct rb_node **rb_link)
{
	node->__rb_parent_color = (unsigned long)parent;
	node->rb_left = node->rb_right = NULL;

	*rb_link = node;
}
```

Reviewer: Those functions could also bring a bit more context:
```cpp
static inline void rb_set_black(struct rb_node *rb)
{
	rb->__rb_parent_color |= RB_BLACK;
}

static inline struct rb_node *rb_red_parent(struct rb_node *red)
{
	return (struct rb_node *)red->__rb_parent_color;
}
```

```

> Therefore, we must set the `struct simple_xattr`'s parent color as `RB_RED` which overlapped with `uaf_chunk`. By doing this, we can control `struct simple_xattr`'s `size` as `start_poll_synchronize_rcu()` return value.

Reviewer: Done

> So, We needs to construct structure's allocation as follows:

Reviewer: ```suggestion
So, we need to construct structure's allocation as follows:
```

```
                             ------------------
                       ╭--> |      parent      | 
                       |     ------------------
                       |                         
              ------------------           ------------------

> uaf_chunk -> |      rchild      |         |      lchild      |

Reviewer: Please add the type to the elements on diagram. I assume that `parent`, `rchild`, `lchild` are all of `simple_xattr`.

Reviewer: Done

              ------------------           ------------------
```

When we modify `uaf_chunk[0][0x20, 0x28)` to `start_poll_synchronize_rcu()` return value, we can change `parent->size` as `start_poll_synchronize_rcu()` return value.

> From [Cross-Cache (per-cpu buddy -> kmalloc-cg-2k)](#cross-cache-per-cpu-buddy---kmalloc-cg-2k), the `lchild`'s size must be use `kmalloc-cg-2k` and we can freely choose `parent`'s slab cache on `kmalloc-cg-64 ~ kmalloc-cg-4k` and `8k, 16k, 32k` managed by buddy allocator.

Reviewer: Done

```cpp
ssize_t
do_getxattr(struct mnt_idmap *idmap, struct dentry *d,
	struct xattr_ctx *ctx)
{
	ssize_t error;
	char *kname = ctx->kname->name;

	if (ctx->size) {
		if (ctx->size > XATTR_SIZE_MAX)
			ctx->size = XATTR_SIZE_MAX;
		ctx->kvalue = kvzalloc(ctx->size, GFP_KERNEL); // [1]
		if (!ctx->kvalue)
			return -ENOMEM;
	}

	if (is_posix_acl_xattr(ctx->kname->name))
		error = do_get_acl(idmap, d, kname, ctx->kvalue, ctx->size);
	else
		error = vfs_getxattr(idmap, d, kname, ctx->kvalue, ctx->size);
	if (error > 0) {
		if (ctx->size && copy_to_user(ctx->value, ctx->kvalue, error)) // [2]
			error = -EFAULT;
	} else if (error == -ERANGE && ctx->size >= XATTR_SIZE_MAX) {
		/* The file system tried to returned a value bigger
		   than XATTR_SIZE_MAX bytes. Not possible. */
		error = -E2BIG;
	}

	return error;
}
```

### OOB Read at `parent` Chunk
#### Leak Controlable kHeap

> From [bypass CONFIG_HARDENED_USERCOPY](#bypass-config_hardened_usercopy), we can read out-of-bounds information of slab cache eventhought `CONFIG_HARDENED_USERCOPY` is enabled.

Reviewer: ```suggestion
From [bypass CONFIG_HARDENED_USERCOPY](#bypass-config_hardened_usercopy), we can read out-of-bounds information of slab cache even though `CONFIG_HARDENED_USERCOPY` is enabled.
```

Reviewer: Done

> WLOG, consider that `parent` is `kmalloc-cg-256` and we spray many `struct simple_xattr`. Then, we can read out-of-bounds information of `kmalloc-cg-256` slab cache. It is shown as follows:

Reviewer: Have we triggered the vulnerability and overwrote the size parameter in parent to have out-of-bounds read? Whiz which value we overwrote it? Please add this to the description.

> WLOG, consider that `parent` is `kmalloc-cg-256` and we spray many `struct simple_xattr`. Then, we can read out-of-bounds information of `kmalloc-cg-256` slab cache. It is shown as follows:

Reviewer: ```suggestion
Without loss of generality, consider that `parent` is `kmalloc-cg-256` and we spray many `struct simple_xattr`. Then, we can read out-of-bounds information of `kmalloc-cg-256` slab cache. It is shown as follows:
```

```
 --------------------------------------
|            parent.rb_node            |
|            parent.name               |
|            parent.size               |
|            parent.value              |
 --------------------------------------
|            xattr<1>.rb_node          |
|            xattr<1>.name             |
|            xattr<1>.size             |
|            xattr<1>.value            |
 --------------------------------------
```

> Therefore, by using OOB read, we can leak `kHeap` address by read `xattr1.rb_node`'s address. The important thing is we can give value at `xattr<*>.value`. Therefore, by reading `xattr<*>.value` and `xattr<*>.rb_node`'s value, we can identify `xattr<*>`'s `lchild`, `rchild`, and `parent`'s kernel address clearly. It dramatically increase the stability of the exploit.

Reviewer: By OOB read you mean calling getxattr() and getting value of xattr? Could you document this in the `exploit.md`.

Reviewer: Done

> Therefore, by using OOB read, we can leak `kHeap` address by read `xattr1.rb_node`'s address. The important thing is we can give value at `xattr<*>.value`. Therefore, by reading `xattr<*>.value` and `xattr<*>.rb_node`'s value, we can identify `xattr<*>`'s `lchild`, `rchild`, and `parent`'s kernel address clearly. It dramatically increase the stability of the exploit.

Reviewer: Do you use this: 

```cpp
#define rb_parent(r)   ((struct rb_node *)((r)->__rb_parent_color & ~3))
```

to calculate the parent's address from `__rb_parent_color`? Please mention it in `exploit.md`.

We can free known kernel address whenever we want. 

#### Leak kbase
When we read OOB data of `parent`, we can read `kbase` address which remained as a uninitialized value.

## Obtain Arbitrary Address Free Primitive
### Cross-Cache (kmalloc-2k -> kmalloc-32k)
By proceeding process at [Trigger Cross-Cache (kmalloc-2k -> per-cpu buddy)](#trigger-cross-cache-kmalloc-2k---per-cpu-buddy), and reallocate to `kmalloc-32k`(not slab cache, but buddy. For convenience, we will call like that). `kmalloc-2k` cache and `kmalloc-32k` cache is same order(order 3), so we can reallocate to `kmalloc-32k` cache.

Then the status of memory is shown as follows:
```
 -------- kmalloc-32k ---------
|             ...              |
|          uaf_chunk           |
|             ...              |
 ------------------------------
```

> uaf_chunk was in `kmalloc-2k` slab cache, so if cross-cached, it will be in somewhere in `kmalloc-32k` slab cache. Then, If we overlap `uaf_chunk` with `struct simple_xattr` which is the size of `kmalloc-32k`, we can modify `uaf_chunk[0]`'s address freely. The reason is that `uaf_chunk[0]` is will be placed on somewhere in `struct simple_xattr.value[]`. 

Reviewer: ```suggestion
uaf_chunk was in `kmalloc-2k` slab cache, so if cross-cached, it will be in somewhere in `kmalloc-32k` slab cache. Then, If we overlap `uaf_chunk` with `struct simple_xattr` which is the size of `kmalloc-32k`, we can modify `uaf_chunk[0]`'s address freely. The reason is that `uaf_chunk[0]` will be placed somewhere in `struct simple_xattr.value[]`. 
```

Let assume that `uaf_chunk[0]`'s address is `xattr<1>.parent - 0x1f` which set at [Cross-Cache (kmalloc-2k -> kmalloc-32k)](#cross-cache-kmalloc-2k---kmalloc-32k). 

Then the status of memory is shown as follows:
```
 -------- uaf_chunk --------                                                offset
|   xattr<1>.parent - 0x1f  | -----> |              ....               |    -0x1f
|            ...            |        |              ....               |    -0x17
 ---------------------------         |              ....               |    -0xf
                                     |              ....               |    -0x7
                                      ------ xattr<1>.parent + 1 ------ 
                                     |              ....               |    +0x1: xattr<1>.parent + 1
                                     |              ....               | 
                                      ---------------------------------
```

> Then, when we trigger [Trigger Vulnerability](#trigger-vulnerability), we can modify `xattr<1>.parent[1, 9)` to `start_poll_synchronize_rcu()` return value. And we know that `start_poll_synchronize_rcu()`'s 7th byte is `0x00`.

Reviewer: Could you elaborate a bit how do we know that `start_poll_synchronize_rcu()`'s 7th byte is `0x00`?

Reviewer: Explained now in `## Primitive` chapter of `exploit.md`

After trigger UAF write and observe memory from the perspective of `struct xattr<1>.parent`. It is shown as follows:
```
 ------------- xattr<1>.parent ------------- 
|  start_poll_synchronize_rcu() value  | ?? |
|       xattr<1> address               | 00 |
|                   ...                     |
 -------------------------------------------
```

> So, we trigger off-by-one at `xattr<1>.parent`. 

Reviewer: Could you explain more what do we modify with off-by-one here? Does it matter for future steps?

Reviewer: Done

### Trigger (Weak) Arbitrary Address Free
Let `xattr<1>.rb_node.parent->rb_right` is allocated in `kmalloc-cg-192`.

#### Why `kmalloc-cg-192`? It was `kmalloc-cg-256`!

> We can make `Let xattr<1>.rb_node.parent->rb_child` by proceeding like below.

Reviewer 2: ```
struct simple_xattr {
	struct rb_node rb_node;
	char * name;
	size_t size;
	char value[];
};

struct rb_node {
	uint64_t  __rb_parent_color;
	struct rb_node *rb_right;
	struct rb_node *rb_left;
};
```
actually, every node must be `xattr<1>.rb_node.parent` instead `xattr<1>.parent`. Do I need to change diagram? It makes diagram too long so I wrote it shortly.

> We can make `Let xattr<1>.rb_node.parent->rb_child` by proceeding like below.

Reviewer: Yes. Please change the diagram to include a full name. It's helps to easier understand the approach.

1. allocate `kmalloc-cg-256`            : `xattr<1>.parent`

> 2. allocate `kmalloc-cg-2k`             : `xattr<1>`

Reviewer: Is this essentially `rchild`, could you add those details here please?

3. leak `xattr<1>.parent`'s address     : from [OOB Read at `parent` Chunk](#oob-read-at-parent-chunk)
4. free `xattr<1>`         

> 5. allocate `kmalloc-cg-192`            : `xattr<1>.rb_node.parent->rb_child`

Reviewer: Do we basically replace object taking `rchild` spot from the one stored in `kmalloc-cg-256` to the one stored in `kmalloc-cg-192`?

Reviewer 2: yes, for control misaligned address fully, we need to change 0x100 -> 0xc0. Usual kernel heap object's size is 2^k, but for exploit only use xattrs, we need 2^k aligned object and not 2^k aligned object.

Reviewer: Please add this to `exploit.md`, it helps easier to understand the logic of exploitation.

Then, we change `kmalloc-cg-256` to `kmalloc-cg-192`.

> For convinience, let's call `rb_child` by `rchild`.

Reviewer: ```suggestion
For convenience, let's call `rb_child` by `rchild`.
```

> Then, when kernel tried to access to `xattr<1>`(`kmalloc-cg-192`), it accessed to `xattr<1> & ~(0xff)` because of [Modify `struct simple_xattr.rb_node->parent.rb_right`](#modify-struct-simple_xattrrb_node-parentrb_right). 

Reviewer: Is it because of off-by-one thing? What parameters with what value we've modified that enabled this behaviour?

Reviewer 2: Yes. If the off by one triggered, `parent.rb_node->rb_child` points other xattr's `xattr.value[]`. so we can fully control all rbtree's element.

> The `kmalloc-cg-192` has 4 cases when off-by-one triggered. It is shown as follows:

Reviewer: Could you describe in more details how do we know that its only 4 cases? Could you maybe draw another diagram to explain this concept in more details?

1. `xattr<1> & 0x3ff` is `0x000`
    - `xattr<1> & ~(0xff)` is same with `xattr<1>`, so nothing changed.
2. `xattr<1> & 0x3ff` is `0x0c0`
    - `xattr<1> & ~(0xff)` is `xattr<1> - 0xc0`, so point other sprayed `&struct simple_xattr.rb_node`.
3. `xattr<1> & 0x3ff` is `0x180`
    - `xattr<1> & ~(0xff)` is `xattr<1> - 0x80`, so point other sprayed `&struct simple_xattr.value[0x18]`.
4. `xattr<1> & 0x3ff` is `0x240`
    - `xattr<1> & ~(0xff)` is `xattr<1> - 0x40`, so point other sprayed `&struct simple_xattr.value[0x58]`.

Therefore, when `xattr<1> & 0x3ff == 0x180` or `xattr<1> & 0x3ff == 0x240`, we can use `simple_xattr.value[]` as fake `struct simple_xattr`.

Therefore, when we set value to `struct simple_xattr.value` as fake chunk, trigger off-by-one, and delete `xattr<1>`, `xattr<1>->name` will be freed. 

Therefore, we can free arbitrary address where we can set `xattr<1>->name`'s prefix as `"security."` (for remove xattr, prefix condition must be satisfied because code find xattr by `xattr<1>->name`).

## Trigger Double Free

> Now, we will trigger double-free on `kmalloc-cg-64` slab cache.

Reviewer 2: pipe_buffer can be allocated only on `2^k` aligned slub object. therefore `kmalloc-cg-64`, `kmalloc-cg-128`, `kmalloc-cg-256` can be used but `kmalloc-cg-192` can't.

> Now, we will trigger double-free on `kmalloc-cg-64` slab cache.

Reviewer: Thank you. Please add this information to `exploit.md`

### Hmm, how we know `kmalloc-cg-64`'s address?
From [OOB Read at `parent` Chunk](#oob-read-at-parent-chunk), we can read out-of-bounds information of slab cache. Therefore, by setting `rchild` as `kmalloc-cg-256`, and `rchild->name` to `kmalloc-cg-64` chunk, we can leak `kmalloc-cg-64`'s address. Also, we can free `kmalloc-cg-64` anytime we want!

Therefore, when we allocate chunk to satisfy Precondition, we can trigger double free
- Precondition: prefix of `&rchild->name[0x28]` is `"security."`(for satisfy [Trigger (Weak) Arbitrary Address Free](#trigger-weak-arbitrary-address-free)) condition

Let's call each chunk as `double_free_rchild` and `double_free_fake_xattr`.

## Arbitrary Address Write & Arbitrary Address Execute
Before AAW, AAE, abstract our primitive.

> - When free `double_free_rchild`, we can free `double_free_chunk`

Reviewer: What is `double_free_chunk`? Where it's allocated?

Reviewer 2: It's just assumption. I just want to explain scenario before real exploit.

- When free `double_free_fake_xattr`, we can free `double_free_chunk + 0x28`

> - `double_free_chunk` is `kmalloc-64`

Reviewer: Is it in `kmalloc-64` or in `kmalloc-cg-64`?

Reviewer 2: Oh, it's `kmalloc-cg-64` not `kmalloc-64`

+0x18   |                                      |
+0x20   |                                      |
                                                ------- double_free_fake_xattr -------
+0x28   |                                      |                                      |
+0x30   |                                      |                                      |
+0x38   |                                      |                                      |
         -------------------------------------- 									  |
+0x40                                          |                                      |
+0x48                                          |                                      |
+0x50                                          |                                      |
+0x58                                          |                                      |
+0x60                                          |                                      |
                                                --------------------------------------     
```           

### Leak VMEMMAP_BASE
Proceding like below, we can leak `VMEMMAP_BASE`.
1. free `double_free_fake_xattr`

> 2. allocate `struct pipe_buffer` at freed `double_free_chunk`

Reviewer: What is the size / cache if `struct pipe_buffer`?

	/* size: 24, cachelines: 1, members: 3 */
	/* last cacheline: 24 bytes */
} __attribute__((__aligned__(8)));
```

Therefore, `uaf_chunk[0]` is `(struct rb_node *)uaf_chunk -> __rb_parent_color`.  `__rb_parent_color` is constructed as follows:
```c++
#define RB_RED          0
#define RB_BLACK        1
#define rb_parent(r)   ((struct rb_node *)((r)->__rb_parent_color & ~3))

static inline void rb_link_node(struct rb_node *node, struct rb_node *parent,
				struct rb_node **rb_link)
{
	node->__rb_parent_color = (unsigned long)parent;
	node->rb_left = node->rb_right = NULL;

	*rb_link = node;
}

static inline void rb_set_black(struct rb_node *rb)
{
	rb->__rb_parent_color |= RB_BLACK;
}

static inline struct rb_node *rb_red_parent(struct rb_node *red)
{
	return (struct rb_node *)red->__rb_parent_color;
}
```

Therefore, we must set the `struct simple_xattr`'s parent color as `RB_RED` which overlapped with `uaf_chunk`. By doing this, we can control `struct simple_xattr`'s `size` as `start_poll_synchronize_rcu()` return value. root node is always `RB_BLACK` following definition of [red-black tree](https://en.wikipedia.org/wiki/Red%E2%80%93black_tree).

> So, We need to construct structure's allocation as follows:

Reviewer: ```suggestion
So we need to construct structure's allocation as follows:
```
